File: Data Structures & Algorithms/valid-tree/submission-2.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        '''
            0 -> 1 -> 2 -> 3 -> 1
            (a, c)
               a    b
              / \  / \
             c   d    e
               a  b   c  d  e
             a -1 -1  0  0 -1
             b -1 -1 -1  0  0
             c -1 -1 -1 -1 -1
             d -1 -1 -1 -1 -1
             e -1 -1 -1 -1 -1

        '''
        if len(edges) <= 0:
            return True
        # Create an adjacency list
        graph = {}
        for i, j in edges:
            if i in graph:
                if j in graph[i]: # if cycle in b/w two adjacent nodes
                    return False
                graph[i].append(j)
            else:
                graph[i] = [j]
            
            if j in graph:
                if i in graph[j]: # if cycle in b/w two adjacent nodes
                    return False
                graph[j].append(i)
            else:
                graph[j] = [i]

        visited = [False for i in graph.keys()]
        
        # just check if all nodes are connected or not and if there is no cycle

        def dfs(node, prev):
            nonlocal graph, visited
            for neighbour in graph[node]:
                if neighbour == prev:
                    continue
                if visited[neighbour] : #cycle
                    return False
                if not visited[neighbour]:
                    visited[neighbour] = True
                    if not (dfs(neighbour, node)):
                        return False
            return True

        visited[0] = True
        if not dfs(0, None):
            return False

        return False if False in visited else True
        

File: Data Structures & Algorithms/max-area-of-island/submission-2.cpp
class Solution {
    int m, n;
public:

    int dfs(int i, int j, vector<vector<int>>& grid, vector<vector<bool>>& visited) {
        if(i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == 0 || visited[i][j] == true) return 0;
        visited[i][j] = true;
        return  1 + dfs(i+1, j, grid, visited) +
                dfs(i-1, j, grid, visited) +
                dfs(i, j-1, grid, visited) + 
                dfs(i, j+1, grid, visited);
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int count = 0;
        m = grid.size(); n = grid[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));

        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(!visited[i][j] && grid[i][j]) {
                    count = max(count, dfs(i, j, grid, visited));
                }
            }
        }
        return count;
    }
};
